### PR TITLE
Static partition forward ranges correctly

### PR DESCRIPTION
**Pull request by BillyONeal**
In DevCom-1326903 the user reports that forward ranges, specifically `std::map` in their case, produce incorrect results from `std::equal`. After reducing to `forward_list` and debugging through, the problem is that when processing "unchunked items" we aren't giving those initial chunks their extra item, so the `std::equal` call fails fast whenever there are unchunked items.

Testing before didn't catch this because the values I chose of being around the thread count didn't account for the underlying infrastructure using `_Oversubscription_multiplier` which meant that the partitioning edge cases were not exercised (in most cases the parallel algorithms tests were testing only 1 element per chunk).

`<execution>` Add the missing +1. Note that this +1 is the same +1 in the iterator-returning version on line 907.

`$/tests/std/include/parallel_algorithms_utilities.hpp`: Adjust the N constants to account for `_Oversubscription_multiplier`.

Resolves DevCom-1326903/VSO-1273957/AB#1273957.

### CODE REVIEW

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

File: tests/std/include/parallel_algorithms_utilities.hpp [hunk 1]
```diff
@@ -7,9 +7,10 @@
 #include <thread>
 
 #ifdef EXHAUSTIVE
-const size_t max_parallel_test_case_n = 1000;
+const size_t max_parallel_test_case_n = 4096;
 #else // ^^^ EXHAUSTIVE ^^^ // vvv !EXHAUSTIVE vvv
-const size_t max_parallel_test_case_n = std::max(4u, std::thread::hardware_concurrency()) + 1;
+// The constant 32 comes from std::_Oversubscription_multiplier
+const size_t max_parallel_test_case_n = std::max(4u, std::thread::hardware_concurrency() * 32) + 1;
```
Comment (StephanTLavavej) [inline tests/std/include/parallel_algorithms_utilities.hpp:13]: We observed `P0024R2_parallel_algorithms_set_difference` and `P0024R2_parallel_algorithms_set_intersection` timeouts internally; I suspect that this is involved.

https://github.com/microsoft/STL/blob/ef6c1ce124cb99115825bfdbf4ed6018ebe8a301/tests/std/include/parallel_algorithms_utilities.hpp#L17-L29

https://github.com/microsoft/STL/blob/ef6c1ce124cb99115825bfdbf4ed6018ebe8a301/tests/std/tests/P0024R2_parallel_algorithms_set_difference/test.cpp#L183

https://github.com/microsoft/STL/blob/ef6c1ce124cb99115825bfdbf4ed6018ebe8a301/tests/std/tests/P0024R2_parallel_algorithms_set_difference/test.cpp#L30-L42

So `testSize` can be 32N. `shortList` contains `testSize / 2` elements, so 16N. Then we iterate for every `overlappingPoint` in `shortList`:

https://github.com/microsoft/STL/blob/ef6c1ce124cb99115825bfdbf4ed6018ebe8a301/tests/std/tests/P0024R2_parallel_algorithms_set_difference/test.cpp#L98-L101

So we're processing 16N elements, 16N times. That's 256 N^2, where N is `thread::hardware_concurrency()`; for N=64 this is 1M.

The tests don't seem to take *too* long on my 32-core machine (256K), but I do notice that they take a while (28 seconds), and my physical machine is very fast. I suspect that with 4x intensity on the slower VMs, we could easily be exceeding the timeouts we set internally (6 minutes), given that each test tries to consume all cores, so if a bunch of other tests are running, very few cores may be effectively available. (In fact, the N^2 pressure means that this gets worse for larger machines; on average each test will have 1 core available because the test runner is already trying to run N tests simultaneously.)

Although we haven't decisively identified the cause of these timeouts, I suspect that something needs to change in this PR.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

**Review by mnatsuhara — APPROVED** [review]

File: tests/std/include/parallel_algorithms_utilities.hpp [hunk 2]
```diff
@@ -7,9 +7,13 @@
 #include <thread>
 
 #ifdef EXHAUSTIVE
-const size_t max_parallel_test_case_n = 1000;
+const size_t max_parallel_test_case_n = 4096;
+// // the number of linear steps a test case that is quadradic should attempt:
```
Comment (StephanTLavavej) [inline tests/std/include/parallel_algorithms_utilities.hpp:11]:
```suggestion
// the number of linear steps a test case that is quadratic should attempt:
```
The variables are also typoed; I'll fix them.

File: tests/std/tests/P0024R2_parallel_algorithms_adjacent_find/test.cpp [hunk 3]
```diff
@@ -33,8 +33,13 @@ void test_case_adjacent_find_parallel(const size_t testSize) {
     assert(less_result == tmp.begin());
     assert(ne_result == tmp.begin());
 
-    auto target = tmp.begin();
+    auto remaining_attempts = quadradic_complexity_case_limit;
+    auto target             = tmp.begin();
     for (auto next = target; ++next != tmp.end(); target = next) {
+        if (--remaining_attempts == 0) {
+            return;
+        }
```
Comment (StephanTLavavej) [inline tests/std/tests/P0024R2_parallel_algorithms_adjacent_find/test.cpp:41]: I observe that there's a counting issue here - for a case limit of L, only L - 1 attempts will be made. The exact number of attempts isn't especially important (and the limit is well above 1), but the variable name doesn't match the control flow. Occurs in most (but not all) files.

Doing something about this would involve a fair number of changes to control flow, so I'm OK with proceeding as-is.

File: tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp [hunk 4]
```diff
@@ -89,9 +118,14 @@ void test_case_is_partitioned_parallel(const size_t testSize) {
...
     // testing with 2 partition adjacent points (T...T F T F...F) where the adjacent partition points are tried at each
     // index
-    first       = c.begin();
-    auto second = next(first, 2);
+    remainingAttempts = quadradic_complexity_case_limit;
+    first             = c.begin();
+    auto second       = next(first, 2);
     for (; second != c.end(); ++first, ++second) {
+        if (--remainingAttempts == 0) {
+            return;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp:126]: This is a minor maintenance hazard. Unlike other tests (where the function contains one small loop), this function contains many loops. Having the last loop use `return;` means that if additional test cases were added below (following the existing pattern of the function), they would be silently skipped.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
FYI @mnatsuhara, I pushed minor changes after you approved.

File: tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp (hunk 4 above)
Comment (BillyONeal) [reply tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp]: I used return here to avoid needing to meet all the postconditions of the loop (iterator positions and so on)

File: tests/std/tests/P0024R2_parallel_algorithms_adjacent_find/test.cpp (hunk 3 above)
Comment (BillyONeal) [reply tests/std/tests/P0024R2_parallel_algorithms_adjacent_find/test.cpp:41]: I think switching to postdecrement instead or moving this if to the end of the loop body would fix it?

File: tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp (hunk 4 above)
Comment (BillyONeal) [reply tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp]: I see you changed this to break but that is not correct because the loop mutates the buffer under test...

File: tests/std/tests/P0024R2_parallel_algorithms_adjacent_find/test.cpp (hunk 3 above)
Comment (BillyONeal) [reply tests/std/tests/P0024R2_parallel_algorithms_adjacent_find/test.cpp]: OK I believe I addressed this

File: tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp (hunk 4 above)
Comment (BillyONeal) [reply tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp]: I guess it doesn't matter for now; followup tests might just clear the buffers or something,

**Review by StephanTLavavej — APPROVED** [review]

File: tests/std/tests/P0024R2_parallel_algorithms_adjacent_find/test.cpp (hunk 3 above)
Comment (StephanTLavavej) [reply tests/std/tests/P0024R2_parallel_algorithms_adjacent_find/test.cpp]: Thanks for going above and beyond! :heart_eyes_cat:

File: tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp (hunk 4 above)
Comment (StephanTLavavej) [reply tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp]: Whoops, sorry for the mistaken change.

**Review by mnatsuhara — APPROVED** [review]

**Comment by CaseyCarter** [comment]
Thanks for this involved fix, Billy-3!

**Comment by BillyONeal** [comment]
> Thanks for this involved fix, Billy-3!

I’ll be sure to leave in lots of typos to get credit for fixing 🥲

At least it isn’t something like, I don’t know, not writing a copy ctor this time

**Comment by CaseyCarter** [comment]
> I’ll be sure to leave in lots of typos to get credit for fixing 🥲

Everyone writes bugs - that's normal. Not everyone is willing to donate their personal time to fix said bugs when it's no longer their day job. That said, there's also a chunk of bonus credit here since it would presumably take someone who didn't originally author the code twice as long to locate the problem and develop a clean fix.
 
> At least it isn’t something like, I don’t know, not writing a copy ctor this time

I'd take your "missing copy constructor" bug over my use-after-free `monotonic_buffer_resource` bug any day of the week.

**Comment by BillyONeal** [comment]
> personal time

I wasn't getting paid for this?

> I'd take your "missing copy constructor" bug over my use-after-free monotonic_buffer_resource bug any day of the week.

Bug off! Bug off! Bug off!....
